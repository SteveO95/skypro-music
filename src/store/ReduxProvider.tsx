'use client';
// Провайдер важный *элемент* который оборачивает приложение и даёт ему доступ к Хранилищу | работать ничего не будет
// Провайдер - это клиентский Компонент (к.к)
import { useRef } from 'react';
import { Provider } from 'react-redux';
import { AppStore, makeStore } from './store';
// Для того что бы вкладывать в К.К серверные компоненты,нужно использовать -children

type ReduxProviderType = {
	children: React.ReactNode;
};

export default function ReduxProvider({ children }: ReduxProviderType) {
	// storeRef- нужен для сохранения переменной между Перерендринга (перерисовок)(что бы наше значение не сбрасывалось)
	const storeRef = useRef<AppStore>();
	if (!storeRef.current) {
		storeRef.current = makeStore();
	}
	// Provider -встроенный компонент(redux.react)в него передаем *Store*-в нём храним (storeRef.current) и children это будет наша страница
	return <Provider store={storeRef.current}>{children}</Provider>;
}
