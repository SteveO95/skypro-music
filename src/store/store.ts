import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { playlistReducer } from "./features/playlistSlice";
import { userReducer } from "./features/userSlice";

export const makeStore = () => {
  return configureStore({
    reducer: combineReducers({
      playlist: playlistReducer,
      user: userReducer,
    }),
  });
};

export type AppStore = ReturnType<typeof makeStore>;
export type RootState = ReturnType<AppStore["getState"]>;
export type AppDispatch = AppStore["dispatch"];
