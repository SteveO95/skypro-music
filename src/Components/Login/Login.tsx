'use client'

import { userApi } from '@/api/userApi'
import { routes } from '@/lib/routes'
import { useAppDispatch } from '@/store/store'
import classNames from 'classnames'
import Image from 'next/image'
import Link from 'next/link'
import { useRouter } from 'next/navigation'
import React, { useMemo, useState } from 'react'
import ErrorMsg from '../Error/ErrorMsg/ErrorMsg'
import styles from './Login.module.css'

export default function Login() {
	const dispatch = useAppDispatch()
	const router = useRouter()

	const [error, setError] = useState<string | null>(null)
	const [loginData, setLoginData] = useState({
		email: '',
		password: '',
	})

	const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const { name, value } = event.target
		setLoginData({ ...loginData, [name]: value })
		setError(null)
	}

	const signIn = useMemo(() => {
		return async (event: React.MouseEvent<HTMLButtonElement>) => {
			event.preventDefault()

			if (!loginData.email.trim() && !loginData.password.trim()) {
				setError('Заполните все поля')
				return
			} else if (!loginData.email.trim()) {
				setError('Введите почту')
				return
			} else if (!loginData.password.trim()) {
				setError('Введите пароль')
				return
			} else if (loginData.password.length < 6) {
				setError('Пароль должен содержать не менее 6 символов.')
				return
			}

			try {
				await dispatch(userApi.getUser(loginData)).unwrap()
				await dispatch(userApi.getToken(loginData))
				router.push(routes.HOME)
			} catch (err) {
				const error = err as Error
				setError(error.message)
				console.error(error.message)
			}
		}
	}, [dispatch, loginData, router])

	return (
		<div className={styles.wrapper}>
			<div className={styles.container_enter}>
				<div className={styles.modal__block}>
					<form className={styles.modal__form_login} action='#'>
						<Link href={routes.HOME}>
							<div className={styles.modal__logo}>
								<Image
									src='/img/logo_modal.png'
									alt='logo'
									width={140}
									height={21}
								/>
							</div>
						</Link>
						<input
							className={classNames(styles.modal__input, styles.login)}
							type='email'
							name='email'
							placeholder='Почта'
							value={loginData.email}
							onChange={onChange}
						/>
						<input
							className={styles.modal__input}
							type='password'
							name='password'
							placeholder='Пароль'
							value={loginData.password}
							onChange={onChange}
							autoComplete='off'
						/>
						{error && <ErrorMsg error={error} />}

						<button className={styles.modal__btn_enter} onClick={signIn}>
							Войти
						</button>
						<button className={styles.modal__btn_signup}>
							<Link href={routes.REGISTER}>Зарегистрироваться</Link>
						</button>
					</form>
				</div>
			</div>
		</div>
	)
}
