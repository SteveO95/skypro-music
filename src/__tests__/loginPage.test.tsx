import Login from '@/components/Login/Login'
import { useAppDispatch } from '@/store/store'
import { useRouter } from 'next/navigation'
import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'

jest.mock('../store/store.ts')
jest.mock('next/navigation')

const mockDispatch = useAppDispatch as jest.Mock
const mockRouter = useRouter as jest.Mock

describe('Страница авторизации', () => {
	beforeEach(() => {
		mockDispatch.mockReturnValue(jest.fn())
		mockRouter.mockReturnValue(jest.fn())
	})

	describe('Рендер элементов на странице', () => {
		it('рендер компонента', () => {
			const login = renderer.create(<Login />).toJSON()
			expect(login).toMatchSnapshot()
		})
		it('лого приложения', () => {
			render(<Login />)

			const logoImg = screen.getByRole('img', { name: 'logo' })

			expect(logoImg).toBeInTheDocument()
		})
		it('поле ввода', () => {
			render(<Login />)

			const input = screen.getByRole('textbox')
			const emailInput = screen.getByPlaceholderText('Почта')
			const passInput = screen.getByPlaceholderText('Пароль')

			expect(input).toBeInTheDocument()
			expect(emailInput).toBeInTheDocument()
			expect(passInput).toBeInTheDocument()
		})
		it('кнопка авторизации и перехода к регистрации', () => {
			render(<Login />)

			const enterBtn = screen.getByRole('button', { name: 'Войти' })
			const regBtn = screen.getByRole('button', { name: 'Зарегистрироваться' })

			expect(enterBtn).toBeInTheDocument()
			expect(regBtn).toBeInTheDocument()
		})
	})

	describe('Валидация полей ввода', () => {
		it('должен выводить сообщение, если все поля ввода пустые', () => {
			render(<Login />)

			const enterBtn = screen.getByText('Войти')

			fireEvent.click(enterBtn)
			expect(screen.getByText('Заполните все поля')).toBeInTheDocument()
		})
		it('должен выводить сообщение, если пароль содержит менее 6 символов', () => {
			render(<Login />)

			const enterBtn = screen.getByText('Войти')
			const emailInput = screen.getByPlaceholderText('Почта')
			const passwordInput = screen.getByPlaceholderText('Пароль')

			fireEvent.change(emailInput, {
				target: {
					value: 'test@mail.com',
				},
			})
			fireEvent.change(passwordInput, {
				target: {
					value: '123',
				},
			})

			fireEvent.click(enterBtn)
			expect(
				screen.getByText('Пароль должен содержать не менее 6 символов.'),
			).toBeInTheDocument()
		})
	})
})
