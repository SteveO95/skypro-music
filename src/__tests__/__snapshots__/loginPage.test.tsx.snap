// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Страница авторизации Рендер элементов на странице рендер компонента 1`] = `
<div
  className="wrapper"
>
  <div
    className="container_enter"
  >
    <div
      className="modal__block"
    >
      <form
        action="#"
        className="modal__form_login"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="modal__logo"
          >
            <img
              alt="logo"
              data-nimg="1"
              decoding="async"
              height={21}
              loading="lazy"
              onError={[Function]}
              onLoad={[Function]}
              src="/_next/image?url=%2Fimg%2Flogo_modal.png&w=384&q=75"
              srcSet="/_next/image?url=%2Fimg%2Flogo_modal.png&w=256&q=75 1x, /_next/image?url=%2Fimg%2Flogo_modal.png&w=384&q=75 2x"
              style={
                {
                  "color": "transparent",
                }
              }
              width={140}
            />
          </div>
        </a>
        <input
          className="modal__input login"
          name="email"
          onChange={[Function]}
          placeholder="Почта"
          type="email"
          value=""
        />
        <input
          autoComplete="off"
          className="modal__input"
          name="password"
          onChange={[Function]}
          placeholder="Пароль"
          type="password"
          value=""
        />
        <button
          className="modal__btn_enter"
          onClick={[Function]}
        >
          Войти
        </button>
        <button
          className="modal__btn_signup"
        >
          <a
            href="/signup"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            Зарегистрироваться
          </a>
        </button>
      </form>
    </div>
  </div>
</div>
`;
